hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java

```diff
@@ -0,0 +1,176 @@
+import com.hazelcast.sql.impl.SqlResultImpl;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Registry of active client cursors.
+ */
+public class QueryClientStateRegistry {
+
+    private final ConcurrentHashMap<QueryId, QueryClientState> clientCursors = new ConcurrentHashMap<>();
+
+    public SqlPage registerAndFetch(
+        UUID clientId,
+        SqlResultImpl cursor,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        QueryClientState clientCursor = new QueryClientState(clientId, cursor);
+
+        SqlPage page = fetchInternal(clientCursor, cursorBufferSize, serializationService);
+
+        if (!page.isLast()) {
+            clientCursors.put(cursor.getQueryId(), clientCursor);
+        }
+
+        return page;
+    }
+
+    public SqlPage fetch(
+        UUID clientId,
+        QueryId queryId,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        QueryClientState clientCursor = getClientCursor(clientId, queryId);
+
+        if (clientCursor == null) {
+            throw QueryException.error("Query cursor is not found (closed?): " + queryId);
+        }
+
+        SqlPage page = fetchInternal(clientCursor, cursorBufferSize, serializationService);
+
+        if (page.isLast()) {
+            deleteClientCursor(clientCursor);
+        }
+
+        return page;
+    }
+
+    private SqlPage fetchInternal(
+        QueryClientState clientCursor,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        Iterator<SqlRow> iterator = clientCursor.getIterator();
+
+        List<SqlPageRow> page = new ArrayList<>(cursorBufferSize);
+        boolean last = fetchPage(iterator, page, cursorBufferSize, serializationService);
+
+        if (last) {
+            deleteClientCursor(clientCursor);
+        }
+
+        return new SqlPage(page, last);
+    }
+
+    private static boolean fetchPage(
+        Iterator<SqlRow> iterator,
+        List<SqlPageRow> page,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        while (iterator.hasNext()) {
+            SqlRow row = iterator.next();
+            SqlPageRow convertedRow = convertRow(row, serializationService);
+
+            page.add(convertedRow);
+
+            if (page.size() == cursorBufferSize) {
+                break;
+            }
+        }
+
+        return !iterator.hasNext();
+    }
+
+    private static SqlPageRow convertRow(SqlRow row, InternalSerializationService serializationService) {
+        int columnCount = row.getMetadata().getColumnCount();
+
+        List<Data> values = new ArrayList<>(columnCount);
+
+        for (int i = 0; i < columnCount; i++) {
+            values.add(serializationService.toData(row.getObject(i)));
```

Comment:
Not sure that I got why do we need to convert the result to `Data` if we are already on the client?

Reply:
This code is invoked on the member

Reply:
Shall we move it then somewhere to the member related class?

Reply:
Makes sense. Moved to the `com.hazelcast.sql.impl.state` package.